cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs

```diff
@@ -0,0 +1,184 @@
+  let strIndex = show index
+  firstExceptT AddressCliError $
+    writeAddress BootstrapAddr (dir </> "address" ++ strIndex ++ ".vkey") addr
+  pure addr
+
+
+-- | Current UTCTime plus 30 seconds
+getCurrentTimePlus30 :: ExceptT CliError IO UTCTime
+getCurrentTimePlus30 =
+    plus30sec <$> liftIO getCurrentTime
+  where
+    plus30sec :: UTCTime -> UTCTime
+    plus30sec = addUTCTime (30 :: NominalDiffTime)
+
+
+readShelleyGenesis :: FilePath -> ExceptT CliError IO (ShelleyGenesis TPraosStandardCrypto)
+readShelleyGenesis fpath = do
+  lbs <- handleIOExceptT (IOError fpath) $ LBS.readFile fpath
+  firstExceptT (AesonDecode fpath . Text.pack) . hoistEither $ Aeson.eitherDecode' lbs
+
+updateTemplate
+    :: SystemStart -> Lovelace
+    -> (Map (GenKeyHash TPraosStandardCrypto) (KeyHash TPraosStandardCrypto)) -> [ShelleyAddress]
+    -> ShelleyGenesis TPraosStandardCrypto -> ShelleyGenesis TPraosStandardCrypto
+updateTemplate start amount delKeys utxoAddrs template =
+    template
+      { sgStartTime = start
+      , sgMaxLovelaceSupply = unsafeGetLovelace amount
+      , sgGenDelegs = delKeys
+      , sgInitialFunds = Map.fromList utxoList
+      }
+  where
+    totalCoin :: Integer
+    totalCoin = fromIntegral (unsafeGetLovelace amount)
+
+    eachAddrCoin :: Integer
+    eachAddrCoin = totalCoin `div` fromIntegral (length utxoAddrs)
+
+    utxoList :: [(Addr TPraosStandardCrypto, Coin)]
+    utxoList = fst $ List.foldl' folder ([], totalCoin) utxoAddrs
+
+    folder :: ([(Addr TPraosStandardCrypto, Coin)], Integer) -> ShelleyAddress
+            -> ([(Addr TPraosStandardCrypto, Coin)], Integer)
+    folder (acc, rest) addr
+      | rest > eachAddrCoin + fromIntegral (length utxoAddrs) = ((addr, Coin eachAddrCoin) : acc, rest - eachAddrCoin)
+      | otherwise = ((addr, Coin rest) : acc, 0)
+
+
+writeShelleyGenesis :: FilePath -> ShelleyGenesis TPraosStandardCrypto -> ExceptT CliError IO ()
+writeShelleyGenesis fpath sg =
+  handleIOExceptT (IOError fpath) $ LBS.writeFile fpath (encodePretty sg)
+
+readGenDelegsMap :: FilePath -> ExceptT CliError IO (Map (GenKeyHash TPraosStandardCrypto) (KeyHash TPraosStandardCrypto))
+readGenDelegsMap gendir = do
+    gkm <- firstExceptT KeyCliError $ readGenesisKeys (gendir </> "genesis-keys")
+    dkm <- firstExceptT KeyCliError $ readDelegateKeys (gendir </> "delegate-keys")
+    -- Both maps should have an identical set of keys.
+    -- The mapMaybe will silently drop any map elements for which the key does not exist
+    -- in both maps.
+    pure $ Map.fromList (mapMaybe (rearrange gkm dkm) $ Map.keys gkm)
+  where
+    rearrange :: Map Int (Ledger.VKey TPraosStandardCrypto) -> Map Int (Ledger.VKey TPraosStandardCrypto)
+             -> Int -> Maybe (GenKeyHash TPraosStandardCrypto, KeyHash TPraosStandardCrypto)
+    rearrange gkm dkm i =
+      case (Map.lookup i gkm, Map.lookup i dkm) of
+        (Just a, Just b) -> Just (coerce (Ledger.hashKey a), Ledger.hashKey b)
```

Comment:
Still need to fix that `coerce`.

Reply:
This will fix the `coerce`
```suggestion
        (Just (Ledger.VKey a), Just b) -> Just (Ledger.hashKey (Ledger.VKeyGenesis a), Ledger.hashKey b)
```